third_party/2and3/yaml/__init__.pyi

```diff
@@ -22,10 +22,27 @@ def safe_load_all(stream: Union[str, IO[str]]) -> Iterator[Any]: ...
 def emit(events, stream=..., Dumper=..., canonical=..., indent=..., width=..., allow_unicode=..., line_break=...): ...
 def serialize_all(nodes, stream=..., Dumper=..., canonical=..., indent=..., width=..., allow_unicode=..., line_break=..., encoding=..., explicit_start=..., explicit_end=..., version=..., tags=...): ...
 def serialize(node, stream=..., Dumper=..., **kwds): ...
-def dump_all(documents: Sequence[Any], stream: IO[str]=..., Dumper=..., default_style=..., default_flow_style=..., canonical=..., indent=..., width=..., allow_unicode=..., line_break=..., encoding=..., explicit_start=..., explicit_end=..., version=..., tags=...) -> Any: ...
-def dump(data: Any, stream: IO[str]=..., Dumper=..., **kwds) -> Any: ...
-def safe_dump_all(documents: Sequence[Any], stream: IO[str]=..., **kwds) -> Any: ...
-def safe_dump(data: Any, stream: IO[str]=..., **kwds) -> Any: ...
+
+@overload
+def dump_all(documents: Sequence[Any], stream: None=..., Dumper=..., default_style=..., default_flow_style=..., canonical=..., indent=..., width=..., allow_unicode=..., line_break=..., encoding=..., explicit_start=..., explicit_end=..., version=..., tags=...) -> AnyStr: ...
+@overload
+def dump_all(documents: Sequence[Any], stream: IO[str]=..., Dumper=..., default_style=..., default_flow_style=..., canonical=..., indent=..., width=..., allow_unicode=..., line_break=..., encoding=..., explicit_start=..., explicit_end=..., version=..., tags=...) -> None: ...
+
+@overload
+def dump(data: Any, stream: None=..., Dumper=..., **kwds) -> AnyStr: ...
```

Comment:
Shouldn't the return be just `str` here and in other functions?

Reply:
Hm, in that case `Union[str, bytes]` may be better. Although I know union return types are strongly discouraged. But the current way of using `AnyStr` as a "hackish" way to implement an unsafe union is purely based on how mypy behaves, it is not specified in PEP 484, so can break at any moment. @JelleZijlstra is this a common practice?

Reply:
I didn't even know this worked like an unsafe union; I've been consistently telling people not to use AnyStr in a single position in a signature.

Regardless, can't this also be fixed with more overloads? It returns unicode if encoding is None or omitted and str when it's a string.

Reply:
You can just add `# type: ignore` if `reveal_type()` shows all correct types all call sites.

Reply:
@ilevkivskyi Your comment confuses me; the `#type` annotation in my last comment is only to document the type returned by `dump()`; that code snippet makes it easy to validate the `yaml/__init__.pyi` file by running `mypy --custom-typeshed-dir=. snipped.py` with and without `--py2` and getting an error message.

So now what:
- [ ] get the `@overload` working to return the correct type depending on `encoding`?
- [ ] replace the `AnyStr` by what?
- [ ] split the `2and3` into two versions to get rid of `Text` being `unicode` with Python2 and `str` with Python3?
- [ ] something other?

Reply:
Add to `# type: ignore` to silence the error you see _in stubs_ (`non-default argument follows default argument`). And then check if this leads to correct results in user code by using `reveal_type()`, this is a mypy trick `reveal_type(<expression>)` creates an error that shows the type of `<expression>` from mypy's point of view.